dashboard/modules/log/log_agent.py

```diff
@@ -20,3 +25,101 @@ async def run(self, server):
     @staticmethod
     def is_minimal_module():
         return False
+
+
+BLOCK_SIZE = 8192
+
+
+class LogAgentV1Grpc(
+    dashboard_utils.DashboardAgentModule, reporter_pb2_grpc.ReporterServiceServicer
+):
+    def __init__(self, dashboard_agent):
+        super().__init__(dashboard_agent)
+
+    async def run(self, server):
+        if server:
+            reporter_pb2_grpc.add_LogServiceServicer_to_server(self, server)
+
+    # TODO: should this return True
+    @staticmethod
+    def is_minimal_module():
+        return False
+
+    async def LogIndex(self, request, context):
+        if os.path.exists(self._dashboard_agent.log_dir):
+            log_files = os.listdir(self._dashboard_agent.log_dir)
+        else:
+            logger.error(
+                f"Could not find log dir at path: {self._dashboard_agent.log_dir}"
+            )
+            log_files = []
+        return reporter_pb2.LogIndexReply(log_files=log_files)
+
+    async def StreamLog(self, request, context):
+        """
+        Streams the log in real time starting from `request.lines` number of lines from
+        the end of the file if `request.keep_alive == True`. Else, it terminates the
+        stream once there are no more bytes to read from the log file.
+        """
+        lines = request.lines if request.lines else 1000
+
+        filepath = f"{self._dashboard_agent.log_dir}/{request.log_file_name}"
+        if not os.path.isfile(filepath):
+            await context.send_initial_metadata(
+                [[log_consts.LOG_STREAM_STATUS, log_consts.FILE_NOT_FOUND]]
+            )
+        else:
+            with open(filepath, "rb") as f:
+                await context.send_initial_metadata(
+                    [[log_consts.LOG_STREAM_STATUS, log_consts.OK]]
+                )
+                # If requesting the whole file, we stream it since it may be large.
+                if lines == -1:
+                    while True:
+                        bytes = f.read(BLOCK_SIZE)
+                        if bytes == b"":
+                            end = f.tell()
+                            break
+                        yield reporter_pb2.StreamLogReply(data=bytes)
```

Comment:
Does it block the event loop or await?

Reply:
If it blocks the event loop, consider adding asyncio.sleep(0) at the end of the while loop